lib/router/src/cache.rs

```diff
@@ -1,426 +1,306 @@
 ///
-/// ## Complexity
+/// Complexity:
+///     - `access` in **O(1)** time (amortized average)
+///     - `insert` in **O(1)** time (average)
 ///
-/// - `access` computes in O(1) time (amortized average).
-/// - `store` computes in O(1) time (average).
-/// - `reserve` computes in O(n) time (average) when capacity is not available,
+/// The underlying data structure of Cache is a [`DelayQueue`]. This allows
+/// the background task to remove values by polling for elements that have
+/// reached their specified deadline. Elements are retrieved from the queue
+/// via [`Stream::poll`], and that is what [`poll_purge`] ultimately uses.
 ///
-/// ### TODO
-///
-/// The underlying datastructure could be improved somewhat so that `reserve` can evict
-/// unused nodes more efficiently. Given that eviction is intended to be rare, this is
-/// probably not a very high priority.
-pub struct Cache<K: Hash + Eq, V, N: Now = ()> {
-    vals: IndexMap<K, Node<V>>,
+/// [`DelayQueue`]: https://docs.rs/tokio/0.1.19/tokio/timer/struct.DelayQueue.html
+/// [`Stream::poll`]: https://docs.rs/tokio/0.1.19/tokio/timer/struct.DelayQueue.html#impl-Stream
+/// [`poll_purge`]: #method.poll_purge
+pub struct Cache<K, V>
+where
+    K: Clone + Eq + Hash,
+{
     capacity: usize,
-    max_idle_age: Duration,
-
-    /// The time source.
-    now: N,
-}
-
-/// Provides the current time within the module. Useful for testing.
-pub trait Now {
-    fn now(&self) -> Instant;
-}
-
-/// Wraps cache values so that each tracks its last access time.
-#[derive(Debug, PartialEq)]
-pub struct Node<T> {
-    value: T,
-    last_access: Instant,
+    expires: Duration,
+    /// Elements are keys into `values`. As elements become Ready, we can
+    /// remove the key and corresponding value from the cache.
+    expirations: DelayQueue<K>,
+    /// Cache access is coordinated through `values`. This field represents
+    /// the current state of the cache.
+    values: IndexMap<K, Node<V>>,
 }
 
-/// A smart pointer that updates an access time when dropped.
+/// A background future that eagerly removes expired cache values.
 ///
-/// Wraps a mutable reference to a `V`-typed value.
-///
-/// When the guard is dropped, the value's `last_access` time is updated with the provided
-/// time source.
-#[derive(Debug)]
-pub struct Access<'a, T: 'a, N: Now + 'a = ()> {
-    node: &'a mut Node<T>,
-    now: &'a N,
-}
-
-/// A handle to a `Cache` that has capacity for at least one additional value.
-#[derive(Debug)]
-pub struct Reserve<'a, K: Hash + Eq + 'a, V: 'a, N: 'a> {
-    vals: &'a mut IndexMap<K, Node<V>>,
-    now: &'a N,
+/// If the cache is dropped, this future will complete.
+pub struct PurgeCache<K, V>
+where
+    K: Clone + Eq + Hash,
+{
+    cache: Lock<Cache<K, V>>,
+    interval: Interval,
 }
 
-#[derive(Clone, Debug, PartialEq)]
-pub struct CapacityExhausted {
-    pub capacity: usize,
+/// A handle to a cache value.
```

Comment:
I don't think I had the Wikipedia definition in mind when describing it, but I use "handle" because it is essentially a wrapper around the actual value so that `dq_key` can be used to do an `expirations` lookup during resets. It's also a private struct so I think handle is fine in this case!